// @stryker-mutator/core/src/utils/Task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const objectUtils_1 = require("./objectUtils");
/**
 * Wraps a promise in a Task api for convenience.
 */
class Task {
    constructor() {
        this._isCompleted = false;
        this.resolve = (result) => {
            this._isCompleted = true;
            this.resolveFn(result);
        };
        this.reject = (reason) => {
            this._isCompleted = true;
            this.rejectFn(reason);
        };
        this._promise = new Promise((resolve, reject) => {
            this.resolveFn = resolve;
            this.rejectFn = reject;
        });
    }
    get promise() {
        return this._promise;
    }
    get isCompleted() {
        return this._isCompleted;
    }
}
exports.Task = Task;
/**
 * A task that can expire after the given time.
 */
class ExpirableTask extends Task {
    constructor(timeoutMS) {
        super();
        this._promise = objectUtils_1.timeout(this._promise, timeoutMS);
    }
}
exports.ExpirableTask = ExpirableTask;
