// @stryker-mutator/mocha-runner/src/StrykerMochaReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_runner_1 = require("@stryker-mutator/api/test_runner");
const Timer_1 = require("./Timer");
class StrykerMochaReporter {
    constructor(runner) {
        this.runner = runner;
        this.timer = new Timer_1.default();
        this.passedCount = 0;
        this.registerEvents();
        StrykerMochaReporter.currentInstance = this;
    }
    registerEvents() {
        this.runner.on('start', () => {
            this.passedCount = 0;
            this.timer.reset();
            this.runResult = {
                errorMessages: [],
                status: test_runner_1.RunStatus.Error,
                tests: []
            };
            StrykerMochaReporter.log.debug('Starting Mocha test run');
        });
        this.runner.on('pass', (test) => {
            this.runResult.tests.push({
                name: test.fullTitle(),
                status: test_runner_1.TestStatus.Success,
                timeSpentMs: this.timer.elapsedMs()
            });
            this.passedCount++;
            this.timer.reset();
        });
        this.runner.on('fail', (test, err) => {
            this.runResult.tests.push({
                failureMessages: [err.message],
                name: test.fullTitle(),
                status: test_runner_1.TestStatus.Failed,
                timeSpentMs: this.timer.elapsedMs()
            });
            if (!this.runResult.errorMessages) {
                this.runResult.errorMessages = [];
            }
            this.runResult.errorMessages.push(err.message);
            if (StrykerMochaReporter.log.isTraceEnabled()) {
                StrykerMochaReporter.log.trace(`Test failed: ${test.fullTitle()}. Error: ${err.message}`);
            }
        });
        this.runner.on('end', () => {
            this.runResult.status = test_runner_1.RunStatus.Complete;
            StrykerMochaReporter.log.debug('Mocha test run completed: %s/%s passed', this.passedCount, this.runResult.tests.length);
        });
    }
}
exports.StrykerMochaReporter = StrykerMochaReporter;
