// @stryker-mutator/core/src/initializer/presets/AngularPreset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const os = require("os");
const handbookUrl = 'https://github.com/stryker-mutator/stryker-handbook/blob/master/stryker/guides/angular.md#angular';
class AngularPreset {
    constructor() {
        this.name = 'angular-cli';
        // Please keep config in sync with handbook
        this.dependencies = [
            '@stryker-mutator/core',
            '@stryker-mutator/karma-runner',
            '@stryker-mutator/typescript',
            '@stryker-mutator/html-reporter'
        ];
        this.config = `{
        mutate: [
          'src/**/*.ts',
          '!src/**/*.spec.ts',
          '!src/test.ts',
          '!src/environments/*.ts'
        ],
        mutator: 'typescript',
        testRunner: 'karma',
        karma: {
          configFile: 'src/karma.conf.js',
          projectType: 'angular-cli',
          config: {
            browsers: ['ChromeHeadless']
          }
        },
        reporters: ['progress', 'clear-text', 'html'],
        maxConcurrentTestRunners: ${Math.floor(os.cpus().length / 2)}, // Recommended to use about half of your available cores when running stryker with angular.
        coverageAnalysis: 'off'
    }`;
    }
    async createConfig() {
        return { config: this.config, handbookUrl, dependencies: this.dependencies };
    }
}
exports.AngularPreset = AngularPreset;
