// @stryker-mutator/core/src/mutants/MutatorFacade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_1 = require("@stryker-mutator/api/plugin");
const di_1 = require("../di");
class MutatorFacade {
    constructor(mutatorDescriptor, pluginCreator, log) {
        this.mutatorDescriptor = mutatorDescriptor;
        this.pluginCreator = pluginCreator;
        this.log = log;
    }
    mutate(inputFiles) {
        const allMutants = this.pluginCreator.create(this.getMutatorName(this.mutatorDescriptor.name)).mutate(inputFiles);
        const includedMutants = this.removeExcludedMutants(allMutants);
        this.logMutantCount(includedMutants.length, allMutants.length);
        return includedMutants;
    }
    removeExcludedMutants(mutants) {
        if (this.mutatorDescriptor.excludedMutations.length) {
            return mutants.filter(mutant => !this.mutatorDescriptor.excludedMutations.includes(mutant.mutatorName));
        }
        else {
            return mutants;
        }
    }
    getMutatorName(mutator) {
        if (typeof mutator === 'string') {
            return mutator;
        }
        else {
            return mutator.name;
        }
    }
    logMutantCount(includedMutantCount, totalMutantCount) {
        let mutantCountMessage;
        if (includedMutantCount) {
            mutantCountMessage = `${includedMutantCount} Mutant(s) generated`;
        }
        else {
            mutantCountMessage = "It's a mutant-free world, nothing to test.";
        }
        const numberExcluded = totalMutantCount - includedMutantCount;
        if (numberExcluded) {
            mutantCountMessage += ` (${numberExcluded} Mutant(s) excluded)`;
        }
        this.log.info(mutantCountMessage);
    }
}
exports.MutatorFacade = MutatorFacade;
MutatorFacade.inject = plugin_1.tokens(plugin_1.commonTokens.mutatorDescriptor, di_1.coreTokens.pluginCreatorMutator, plugin_1.commonTokens.logger);
