// @stryker-mutator/api/src/config/Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("../../core");
const DashboardOptions_1 = require("../core/DashboardOptions");
const DefaultTempDirName_1 = require("./DefaultTempDirName");
class Config {
    constructor() {
        this.mutate = [
            '{src,lib}/**/*.js?(x)',
            '!{src,lib}/**/__tests__/**/*.js?(x)',
            '!{src,lib}/**/?(*.)+(spec|test).js?(x)',
            '!{src,lib}/**/*+(Spec|Test).js?(x)'
        ];
        this.logLevel = core_1.LogLevel.Information;
        this.fileLogLevel = core_1.LogLevel.Off;
        this.timeoutMS = 5000;
        this.timeoutFactor = 1.5;
        this.plugins = ['@stryker-mutator/*'];
        this.reporters = ['progress', 'clear-text'];
        this.coverageAnalysis = 'off';
        this.testRunner = 'command';
        this.mutator = 'javascript';
        this.transpilers = [];
        this.maxConcurrentTestRunners = Infinity;
        this.symlinkNodeModules = true;
        this.thresholds = {
            break: null,
            high: 80,
            low: 60
        };
        this.allowConsoleColors = true;
        /**
         * The options for the 'dashboard' reporter
         */
        this.dashboard = {
            baseUrl: 'https://dashboard.stryker-mutator.io/api/reports',
            reportType: DashboardOptions_1.ReportType.MutationScore
        };
        this.tempDirName = DefaultTempDirName_1.default;
    }
    set(newConfig) {
        if (newConfig) {
            Object.keys(newConfig).forEach(key => {
                if (newConfig[key] !== undefined) {
                    if (key === 'dashboard') {
                        this[key] = Object.assign(Object.assign({}, this[key]), newConfig[key]);
                    }
                    else {
                        this[key] = newConfig[key];
                    }
                }
            });
        }
    }
}
exports.default = Config;
