// @stryker-mutator/jest-runner/src/configLoaders/ReactScriptsTSJestConfigLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("path"));
const createReactJestConfig_1 = require("../utils/createReactJestConfig");
class ReactScriptsTSJestConfigLoader {
    constructor(projectRoot, resolve = require.resolve) {
        this.resolve = resolve;
        this.projectRoot = projectRoot;
    }
    loadConfig() {
        try {
            // Get the location of react-ts script, this is later used to generate the Jest configuration used for React projects.
            const reactScriptsTsLocation = path.join(this.resolve('react-scripts-ts/package.json'), '..');
            // Create the React configuration for Jest
            const jestConfiguration = this.createJestConfig(reactScriptsTsLocation);
            // Set test environment to jsdom (otherwise Jest won't run)
            jestConfiguration.testEnvironment = 'jsdom';
            return jestConfiguration;
        }
        catch (e) {
            if (this.isNodeErrnoException(e) && e.code === 'MODULE_NOT_FOUND') {
                throw Error('Unable to locate package react-scripts-ts. ' + 'This package is required when projectType is set to "react-ts".');
            }
            throw e;
        }
    }
    isNodeErrnoException(arg) {
        return arg.code !== undefined;
    }
    createJestConfig(reactScriptsTsLocation) {
        return createReactJestConfig_1.createReactTsJestConfig((relativePath) => path.join(reactScriptsTsLocation, relativePath), this.projectRoot, false);
    }
}
exports.default = ReactScriptsTSJestConfigLoader;
