// @stryker-mutator/jest-runner/src/configLoaders/ReactScriptsJestConfigLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const createReactJestConfig_1 = require("../utils/createReactJestConfig");
class ReactScriptsJestConfigLoader {
    constructor(projectRoot, resolve = require.resolve) {
        this.resolve = resolve;
        this.projectRoot = projectRoot;
    }
    loadConfig() {
        try {
            // Get the location of react script, this is later used to generate the Jest configuration used for React projects.
            const reactScriptsLocation = path_1.default.join(this.resolve('react-scripts/package.json'), '..');
            // Create the React configuration for Jest
            const jestConfiguration = this.createJestConfig(reactScriptsLocation);
            // Set test environment to jsdom (otherwise Jest won't run)
            jestConfiguration.testEnvironment = 'jsdom';
            return jestConfiguration;
        }
        catch (e) {
            if (this.isNodeErrnoException(e) && e.code === 'MODULE_NOT_FOUND') {
                throw Error('Unable to locate package react-scripts. This package is required when projectType is set to "react".');
            }
            throw e;
        }
    }
    isNodeErrnoException(arg) {
        return arg.code !== undefined;
    }
    createJestConfig(reactScriptsLocation) {
        return createReactJestConfig_1.createReactJestConfig((relativePath) => path_1.default.join(reactScriptsLocation, relativePath), this.projectRoot, false);
    }
}
exports.default = ReactScriptsJestConfigLoader;
