// @stryker-mutator/core/src/di/coreTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cliOptions = 'cliOptions';
exports.configReader = 'configReader';
exports.configReadFromConfigFile = 'configReadFromConfigFile';
exports.configEditorApplier = 'configEditorApplier';
exports.inputFiles = 'inputFiles';
exports.initialRunResult = 'initialRunResult';
exports.transpiledFiles = 'transpiledFiles';
exports.mutantTranspileScheduler = 'mutantTranspileScheduler';
exports.sandboxPool = 'sandboxPool';
exports.temporaryDirectory = 'temporaryDirectory';
exports.testFramework = 'testFramework';
exports.timer = 'timer';
exports.timeOverheadMS = 'timeOverheadMS';
exports.loggingContext = 'loggingContext';
exports.transpiler = 'transpiler';
exports.sandboxIndex = 'sandboxIndex';
exports.reporter = 'reporter';
exports.pluginKind = 'pluginKind';
exports.pluginDescriptors = 'pluginDescriptors';
exports.pluginCreatorReporter = 'pluginCreatorReporter';
exports.pluginCreatorConfigEditor = 'pluginCreatorConfigEditor';
exports.pluginCreatorTranspiler = 'pluginCreatorTranspiler';
exports.pluginCreatorTestRunner = 'pluginCreatorTestRunner';
exports.pluginCreatorMutator = 'pluginCreatorMutator';
exports.pluginCreatorTestFramework = 'pluginCreatorTestFramework';
