// @stryker-mutator/core/src/di/factoryMethods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_1 = require("@stryker-mutator/api/plugin");
const TestFrameworkOrchestrator_1 = require("../TestFrameworkOrchestrator");
const objectUtils_1 = require("../utils/objectUtils");
const _1 = require(".");
function pluginResolverFactory(injector) {
    const pluginLoader = injector.injectClass(_1.PluginLoader);
    pluginLoader.load();
    return pluginLoader;
}
exports.pluginResolverFactory = pluginResolverFactory;
pluginResolverFactory.inject = plugin_1.tokens(plugin_1.commonTokens.injector);
function testFrameworkFactory(injector) {
    return injector.injectClass(TestFrameworkOrchestrator_1.default).determineTestFramework();
}
exports.testFrameworkFactory = testFrameworkFactory;
testFrameworkFactory.inject = plugin_1.tokens(plugin_1.commonTokens.injector);
function loggerFactory(getLogger, target) {
    return getLogger(target ? target.name : 'UNKNOWN');
}
exports.loggerFactory = loggerFactory;
loggerFactory.inject = plugin_1.tokens(plugin_1.commonTokens.getLogger, plugin_1.commonTokens.target);
function optionsFactory(config, configEditorApplier) {
    configEditorApplier.edit(config);
    return objectUtils_1.freezeRecursively(config);
}
exports.optionsFactory = optionsFactory;
optionsFactory.inject = plugin_1.tokens(_1.coreTokens.configReadFromConfigFile, _1.coreTokens.configEditorApplier);
function mutatorDescriptorFactory(options) {
    const defaults = {
        plugins: null,
        name: 'javascript',
        excludedMutations: []
    };
    if (typeof options.mutator === 'string') {
        return Object.assign(Object.assign({}, defaults), { name: options.mutator });
    }
    return Object.assign(Object.assign({}, defaults), options.mutator);
}
exports.mutatorDescriptorFactory = mutatorDescriptorFactory;
mutatorDescriptorFactory.inject = plugin_1.tokens(plugin_1.commonTokens.options);
