// @stryker-mutator/core/src/process/MutationTestExecutor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_1 = require("@stryker-mutator/api/plugin");
const operators_1 = require("rxjs/operators");
const di_1 = require("../di");
class MutationTestExecutor {
    constructor(reporter, mutantTranspileScheduler, sandboxPool) {
        this.reporter = reporter;
        this.mutantTranspileScheduler = mutantTranspileScheduler;
        this.sandboxPool = sandboxPool;
        this.reportResult = (mutantResult) => {
            this.reporter.onMutantTested(mutantResult);
        };
        this.reportAll = (mutantResults) => {
            this.reporter.onAllMutantsTested(mutantResults);
        };
    }
    async run(allMutants) {
        const results = await this.sandboxPool
            .runMutants(this.mutantTranspileScheduler.scheduleTranspileMutants(allMutants))
            .pipe(operators_1.tap(this.reportResult), 
        // Signal the mutant transpiler that there is another slot open for transpiling
        operators_1.tap(this.mutantTranspileScheduler.scheduleNext), operators_1.toArray(), operators_1.tap(this.reportAll))
            .toPromise();
        return results;
    }
}
exports.MutationTestExecutor = MutationTestExecutor;
MutationTestExecutor.inject = plugin_1.tokens(di_1.coreTokens.reporter, di_1.coreTokens.mutantTranspileScheduler, di_1.coreTokens.sandboxPool);
